src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -453,14 +475,12 @@ public SolutionState RemoveProjects(ArrayBuilder<ProjectId> projectIds)
 
         var languageCountDeltas = new TemporaryArray<(string language, int count)>();
         foreach (var projectId in projectIds)
-        {
-            AddLanguageCountDelta(ref languageCountDeltas, ProjectStates[projectId].Language, amount: -1);
-        }
+            AddLanguageCountDelta(ref languageCountDeltas, GetProjectState(projectId)!.Language, amount: -1);
 
         return this.Branch(
             solutionAttributes: newSolutionAttributes,
             projectIds: newProjectIds,
-            idToProjectStateMap: newStateMap,
+            projectStates: newProjectStates,
             projectCountByLanguage: AddLanguageCounts(ProjectCountByLanguage, languageCountDeltas),
```

Comment:
Wait, we keep a map of this? That'd be useful for your other PR about updating UI contexts where we're just enumerating all projects again...

Reply:
I'm not too worried about that enumeration, plus the comments in that code indicated that it's trying not to read CurrentSolution off the main thread.